### PR TITLE
add circular dependency detection, dep. tree, injector name

### PR DESCRIPTION
**Pull request by lukseven**

### CODE REVIEW

File: context.go [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+)
+
+// ctx is the dependency resolution context. It is used to detect circular
+// dependencies and provide the dependency tree.
+type ctx struct {
+	root    *stack
+	current *stack
+}
+
+func zeroCtx() ctx {
```
Comment (jslching) [inline context.go]: Looks like this isn't used anywhere? Also, just curious, what is the reason for making it a function, rather than just a variable?

**Review by jslching — APPROVED** [review]
LGTM

File: context.go [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+		itype = "child"
+	}
+	root := newStack(nil, rootBindingKey{itype}, rootBinding{inj.name})
+	return ctx{root: root, current: root}
+}
+
+func (c *ctx) push(key bindingKey, binding resolvedBinding) (err error) {
+	if c.root == nil {
+		return nil
+	}
```
Comment (elv-gilles) [inline context.go]: I don't see a place where c.root can be nil

**Review by elv-gilles — APPROVED** [review]
The dependency tree of qfab looks nice .. and huge.
Sadly, there's no way to gain access to the dependency tree of children injectors.

**Comment by lukseven** [comment]
> The dependency tree of qfab looks nice .. and huge. Sadly, there's no way to gain access to the dependency tree of children injectors.

A root injector knows nothing of a child injector. You can get the dependency tree where the child injector is created. See example app: https://github.com/eluv-io/inject-go/pull/3/files#diff-2cd8946462e821ac21389b67c0bcde485749ba0ab0cfa12dd5b2b3c50b5591e0

File: context.go (hunk 1 above)
Comment (lukseven) [reply context.go]: Right, not used anymore in final version. Will remove. 
A function guarantees that nobody can change the value, which is not the case with global variables. Also, the function will be inlined by the compiler, so is equivalent to actually creating a zero value in the calling code (i.e. `ctx{}`).

File: context.go (hunk 2 above)
Comment (lukseven) [reply context.go]: zeroCtx()

**Pull request merged**